### PR TITLE
scanvi updates

### PR DESCRIPTION
**Pull request by galenxing**

### CODE REVIEW

File: scvi/dataloaders/_concat_dataloader.py [hunk 1]
```diff
@@ -40,3 +40,69 @@ def __iter__(self):
...
+            batch_size=batch_size,
+            data_and_attributes=data_and_attributes,
+            **data_loader_kwargs,
+        )
+
+    def resample_labels(self):
+        labelled_idx = self.subsample_labels()
+        # self.dataloaders[0] iterates over full_indices
+        # self.dataloaders[1] iterates over the labelled_indices
+        # change the indicees of the labelled set
```
Comment (adamgayoso) [inline scvi/dataloaders/_concat_dataloader.py:91]:
```suggestion
        # change the indices of the labelled set
```

File: scvi/dataloaders/_concat_dataloader.py [hunk 2]
```diff
@@ -40,3 +40,69 @@ def __iter__(self):
...
+        labels_obs_key,
+        unlabeled_category,
+        n_samples_per_label,
+        indices=None,
+        shuffle=False,
+        batch_size=128,
+        data_and_attributes: Optional[dict] = None,
+        **data_loader_kwargs,
+    ):
+        if indices is None:
```
Comment (adamgayoso) [inline scvi/dataloaders/_concat_dataloader.py:58]: docstring with general capabilities and document args, typing in args

File: scvi/dataloaders/_concat_dataloader.py [hunk 3]
```diff
@@ -40,3 +40,69 @@ def __iter__(self):
...
+        )
+
+    def resample_labels(self):
+        labelled_idx = self.subsample_labels()
+        # self.dataloaders[0] iterates over full_indices
+        # self.dataloaders[1] iterates over the labelled_indices
+        # change the indicees of the labelled set
+        self.dataloaders[1].indices = labelled_idx
+
+    def subsample_labels(self):
```
Comment (adamgayoso) [inline scvi/dataloaders/_concat_dataloader.py:94]: docstring

File: scvi/lightning/_sampling.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from pytorch_lightning.callbacks import Callback
```
Comment (adamgayoso) [inline scvi/lightning/_sampling.py:1]: should we for now have a file just called `_callbacks.py` with this in it?

File: tests/models/test_models.py [hunk 5]
```diff
@@ -233,12 +235,38 @@ def test_save_load_model(cls, adata, save_path):
     assert model.is_trained is True
 
 
+def test_SemiSupervisedDataLoader():
```
Comment (adamgayoso) [inline tests/models/test_models.py:238]:
```suggestion
def test_semisupervised_dataloader():
```

File: scvi/model/_scanvi.py [hunk 6]
```diff
@@ -75,45 +76,34 @@ def __init__(
...
-        pretrained_model: Optional[SCVI] = None,
         n_hidden: int = 128,
         n_latent: int = 10,
         n_layers: int = 1,
         dropout_rate: float = 0.1,
         dispersion: Literal["gene", "gene-batch", "gene-label", "gene-cell"] = "gene",
         gene_likelihood: Literal["zinb", "nb", "poisson"] = "zinb",
         use_gpu: bool = True,
-        **model_kwargs,
+        encode_covariates: bool = False,
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:86]: shouldn't be necessary now

File: scvi/model/_scanvi.py [hunk 7]
```diff
@@ -75,45 +76,34 @@ def __init__(
...
+        scanvae_model_kwargs: dict = {},
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:87]:
```suggestion
        **model_kwargs,
```
and make sure to update docstring

File: scvi/model/_scanvi.py [hunk 8]
```diff
@@ -198,37 +191,31 @@ def _data_loader_cls(self):
...
-            "semisupervised_trainer_history": self.semisupervised_history_,
-        }
+        return self._trainer.logger.history
 
     def train(
         self,
-        n_epochs_unsupervised: Optional[int] = None,
-        n_epochs_semisupervised: Optional[int] = None,
+        max_epochs: Optional[int] = None,
+        n_samples_per_label: Optional[float] = None,
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:199]: needs to be in docstring

File: scvi/model/_scanvi.py [hunk 9]
```diff
@@ -198,37 +191,31 @@ def _data_loader_cls(self):
...
         Train the model.
 
         Parameters
         ----------
-        n_epochs_unsupervised
-            Number of passes through the dataset for unsupervised pre-training.
-        n_epochs_semisupervised
+        max_epochs
             Number of passes through the dataset for semisupervised training.
         train_base_model
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:220]: should be removed

File: scvi/model/_scanvi.py [hunk 10]
```diff
@@ -198,37 +191,31 @@ def _data_loader_cls(self):
...
 
     def train(
         self,
-        n_epochs_unsupervised: Optional[int] = None,
-        n_epochs_semisupervised: Optional[int] = None,
+        max_epochs: Optional[int] = None,
+        n_samples_per_label: Optional[float] = None,
         train_size: float = 0.9,
         validation_size: Optional[float] = None,
         batch_size: int = 512,
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:202]: 512 default? Also missing from docstring

File: scvi/model/_scanvi.py [hunk 11]
```diff
@@ -198,37 +191,31 @@ def _data_loader_cls(self):
...
         use_gpu: Optional[bool] = None,
+        num_workers: int = 4,
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:204]: idk how I feel about num_workers being accessible here. I feel like it could be good if we (for the next PR)

1. Add `dl_num_workers` to our settings config with the pytorch default.
2. Catch the pytorch lightning user warning about it and create a more informative logger warning about it, as well as how to change it.
3. Add `pin_memory_if_gpu = True` to our settings.

Basically batch size is more of a training parameter like learning rate than the dataloader kwargs, and those things should be isolated IMO

Comment (adamgayoso) [reply scvi/model/_scanvi.py:204]: I added this to notion. For now, let's remove num_workers

File: scvi/model/_scanvi.py [hunk 12]
```diff
@@ -198,37 +191,31 @@ def _data_loader_cls(self):
...
         train_base_model: bool = True,
         lr: float = 1e-3,
         n_epochs_kl_warmup: int = 400,
         n_iter_kl_warmup: Optional[int] = None,
         check_val_every_n_epoch: Optional[int] = None,
-        unsupervised_trainer_kwargs: dict = {},
-        semisupervised_trainer_kwargs: dict = {},
-        unsupervised_task_kwargs: dict = {},
-        semisupervised_task_kwargs: dict = {},
+        trainer_kwargs: dict = {},
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:210]: these shouldn't be {} by default. Please take a look at the totalVI train method. I exposed certain important things.

File: scvi/model/_scanvi.py [hunk 13]
```diff
@@ -376,8 +344,141 @@ def predict(
...
+        self,
+        adata: AnnData,
+        unlabeled_category,
+        labels_obs_key,
+        train_size: float = 0.9,
+        validation_size: Optional[float] = None,
+        batch_size: Optional[int] = None,
+        n_samples_per_label=100,
+        pin_memory: bool = False,
+        num_workers: int = 4,
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:363]: remove num workers (for now)

**Review by adamgayoso — CHANGES_REQUESTED** [review]
